data_mgr.py

```diff
@@ -0,0 +1,296 @@
+                result.append(node)
+        return result
+
+    def get_node_id(self, node_type, args={}):
+        """Return GrapphQL task object for given id."""
+        nat_id = args.get('native_ids', None)
+        oname, wname, nid = nat_id.split('/', 2)
+        w_id = f"{oname}/{wname}"
+        if node_type == 'proxy_nodes':
+            nodes = (
+                list(getattr(self.data[w_id], 'task_proxies', []))
+                + list(getattr(self.data[w_id], 'family_proxies', [])))
+        else:
+            nodes = getattr(self.data[w_id], node_type, [])
+        for node in nodes:
+            if node.id == nat_id:
+                return node
+        return None
+
+    def get_edges_all(self, args={}):
+        w_args = {
+            'ids': args['workflows'],
+            'exids': args['exworkflows'],
+        }
+        result = []
+        for edge in [
+                n for k in self.get_workflow_msgs(w_args)
+                for n in getattr(k, 'edges')]:
+            result.append(edge)
+        return result
+
+    def get_edges_id(self, args={}):
+        nat_ids = args.get('native_ids', [])
+        w_ids = []
+        for nat_id in nat_ids:
+            oname, wname, eid = nat_id.split('/', 2)
+            w_ids.append(f"{oname}/{wname}")
+        w_ids = list(set(w_ids))
+        result = []
+        for edge in [
+                e for k in w_ids
+                for e in getattr(self.data[k], 'edges')]:
+            result.append(edge)
+        return result
+
+    def _node_filter(self, node, ids=[], exids=[], args={}):
+        natts = [
+            getattr(node, 'cycle_point', None),
+            getattr(node, 'namespace', [
+                n for n in [getattr(node, 'name', None)] if n is not None]),
+            getattr(node, 'submit_num', None),
+            getattr(node, 'state', None),
+        ]
+        if (
+                (not args.get('states') or node.state in args['states']) and
+                not (
+                    args.get('exstates') and
+                    node.state in args['exstates']) and
+                (args.get('mindepth', -1) < 0 or
+                    node.depth >= args['mindepth']) and
+                (args.get('maxdepth', -1) < 0 or
+                    node.depth <= args['maxdepth']) and
+                (not ids or self._node_atts_filter(natts, ids)) and
+                not (exids and self._node_atts_filter(natts, exids))):
+            return True
+        return False
+
+    @staticmethod
+    def _node_atts_filter(natts, items):
+        for point, name, submit_num, state in items:
+            if ((not point or fnmatchcase(natts[0], point)) and
+                    any(fnmatchcase(nn, name) for nn in natts[1]) and
+                    (not submit_num or fnmatchcase(natts[2], submit_num)) and
+                    (not state or natts[3] == state)):
+                return True
+        return False
+
+    def _workflow_filter(self, flow, ids=[], exids=[], args={}):
```

Comment:
Mutable default arguments. And `args` is never used?

Reply:
It's not used yet, however as soon as I add filters for filtering workflows by status or something, then it will be.. (will remove for now)